### DISCUSSION TITLE
VcDatasourceXXX组件的errorEvent事件未触发。

### DISCUSSION THREAD

**Discussion by houluy** [zouyaoji/vue-cesium#260] [Q&A] [answered]
作者大大，您好！

我利用VcDatasourceGeojson和kml组件加载了一些Geoserver的矢量。我在服务端加了个鉴权，对请求返回401错误码，想在前端通过errorEvent事件捕捉这个错误码进行处理，但是貌似errorEvent并未触发，直接内部报了个Get请求401未授权的错误。请问是不是我使用方式有误？

```vue
<template>
<VcDatasourceGeojson :data="item" v-for="(item, index) in geojson" :key="index" @errorEvent="loadError"></VcDatasourceGeojson>
</template>
<script setup>
let geojson = reactive(["http://xxxxx/ows?xxx"])
const loadError = () => {
  console.log("ERROR") // 这个ERROR并未打印出来
}
</script>
```

浏览器显示的错误图片如下，似乎在Cesium中抛出了错误，但并没有进行捕捉：

![Inkederror](https://user-images.githubusercontent.com/11565649/185835905-102d3023-3573-47d4-921d-ad83310a4abb.jpg)

谢谢作者大大！

**Comment by zouyaoji** [comment, marked as answer]
这个只有你写个复现例子我测测看看了。或者你自己用 GeoJsonDataSource 试试会不会触发这儿。
https://github.com/CesiumGS/cesium/blob/1.96/Source/DataSources/GeoJsonDataSource.js#L967

**Reply by houluy** [reply]
我试了一下，用Cesium原生GeoJsonDataSource是会throw出err，然后通过catch能够捕捉到。如果用vue-cesium的VcDatasourceGeojson组件的error-event事件，不会触发。如下是测试程序：分别用Cesium和VcDatasourceGeojson访问同一个url（服务器会响应401未授权错误）。

```vue
<script lang="ts" setup>
import { ref } from 'vue'
const cesiumPath = ref("Cesium/Cesium.js")
// 实际的域名和端口号会发您邮箱
let dsUrl = ref("http://localhost/test/geoserver/Cesium/ows?service=WFS&version=1.0.0&request=GetFeature&typeName=Cesium%3ABeijing&maxFeatures=50&outputFormat=application%2Fjson")
const onViewerReady = (readyObj) => {
  let url = dsUrl.value
  readyObj.Cesium.GeoJsonDataSource.load(url)
  .catch(err => {
    console.log("Error!", err)
  })
}
const onErrorEvent = () => {
  console.log("on Error Event")
}
</script>
<template>
<vc-viewer :cesiumPath="cesiumPath" @ready="onViewerReady">
  <VcDatasourceGeojson :data="dsUrl" @error-event="onErrorEvent"></VcDatasourceGeojson>
</vc-viewer>
</template>
```

输出结果：

![Inkederror截图](https://user-images.githubusercontent.com/11565649/185896285-1dba608f-9984-4d2c-9d74-2ed441c45d01.jpg)

**Reply by zouyaoji** [reply]
已经确认是 vue-cesium 的bug。 但这个bug也是vue-cesium的软肋，不太好解决，vue-cesium注册事件的逻辑，说是注册，但转发更为准确。就是要在cesium对象创建成功之后，才能转发事件。

https://github.com/zouyaoji/vue-cesium/blob/dev/packages/composables/use-common/index.ts#L151
https://github.com/zouyaoji/vue-cesium/blob/dev/packages/composables/use-events/index.ts#L22

咱们现在的 GeojsonDatasouece 是在创建过程中遇到了错误，所以事件实际上都没注册上呢。


解决办法
1、通用办法：vue-cesium 为所有组件的增加一个通用的，创建对象遇到错误然后抛出异常的事件，如有创建过程错误中遇到错误，用这个事件抛出异常。这个事件取名 unready，下个版本给出。
2、针对VcDatasourceGeojson的话，可以由您这边自己处理。VcDatasourceGeojson的data参数不仅支持传url，还可以直接传对象。这个对象由你外部请求成功之后再赋值给data。这样你就可以自己处理了。

**Reply by zouyaoji** [reply]
![image](https://user-images.githubusercontent.com/24430116/185917693-a5433ae0-a916-4b95-802c-56d2482f532a.png)

![image](https://user-images.githubusercontent.com/24430116/185917618-5102094a-cff4-47a6-a5d9-16ce57a008eb.png)

**Reply by houluy** [reply]
Ok，明白了。那我们先用第二种方法自己去获取数据再加载组件。
谢谢~

**Reply by zouyaoji** [reply]
预计明天晚上发版本了就有 unready 事件。

**Reply by zouyaoji** [reply]
代码已经提交，有兴趣可以自己拉代码，本地打包，将打包结果的dist中的vue-cesium替换你项目node_modules中的vue-cesium即可测试。

**Reply by houluy** [reply]
万分感谢~~
